effect/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -39,31 +43,31 @@ import scalaz.data.Maybe
  * values, see the default interpreter in `RTS` or the safe main function in
  * `SafeApp`.
  */
-sealed abstract class IO[A] { self =>
+sealed abstract class IO[E, A] { self =>
```

Comment:
Not really sure what it is that we expect `E` to be. From my point of view, `E` can't really be anything but some `E <: Exception`. If we are considering `E` to be any error type, like `String` for example, then how is that different than `EitherT[IO, E, A]`? (where `IO` is the current IO). 

Now, one cool thing that can happen from exposing this `E` is that we can now have `IO[Exception with NoStackTrace, A]`. This will allow your code to use exceptions without the stacktrace issue (and even control it). Not sure how impure code will convert from their Exception/Throwable into a NoStrackTrace one, but maybe there's way.

Reply:
There is no requirement on `E`. It can be some ADT, unrelated to exceptions, or even just `String` if all you care about is error messages; `Void` / `Nothing` if you don't want any errors; or `Unit` if you want to treat an IO as `F[Option[A]]`. 

You can `leftMap` over `E` to "lift" an IO from one error type to another, which lets you change your error types in different parts of your application.

Reply:
Then I'm confused: what's the difference between `IO[E, A]` and `EitherT[IO, E, A]`?

Reply:
> what's the difference between IO[E, A] and EitherT[IO, E, A]

considering an `IO[A]` that never catch exceptions, there are isomorphic (I think), the difference is better performance and ease of use (no/less monad transformers).

> and how does E interface with flat out thrown exceptions?

See [`partialSnyc`](https://github.com/jdegoes/scalaz/blob/a05f9f1f0ecde46159b598d92f19f9de5e56d22c/effect/shared/src/main/scala/scalaz/effect/IO.scala#L638). The strategy here is to deffer constraints on type parameters until you actually need them, to maximize possible usages. It is similar to how one does not need an `Order[A]` to construct a `ISet.empty[A]`.

Reply:
@pjrt 

The type `EitherT[IO, E, A]` — for Scalaz 7, Monix Task, Cats Effect, Future, etc. — has two separate error channels, and two `MonadError` instances (one is fixed to `Throwable`, and one varies as `E`).

You can and probably should think of `EitherT[IO, E, A]` for these effect monads as being equivalent to `EitherT[EitherT[Unexceptional, Throwable, ?], E, A]` (for some unexceptional effect monad `Unexceptional`), which is a nesting of 2 layers of `Either`. Such values may "fail" — in the functional programming sense — and be recovered from in two distinct ways: the inner `Either` (with `Throwable`), or the outer `Either` (with `E`).

In contrast, `IO[E, A]` has a single `MonadError` instance, and should be thought of as equivalent to `EitherT[Unexceptional, E, A]`. There is only one way it may "fail" — in the functional programming sense — and be recovered from: with `E`.

With the older effect monads, because there are two separate error channels, and because one of them is fixed at compile-time to `Throwable`, it limits the precision with which you can describe effects. You cannot describe effects that cannot fail, nor can you describe effects that can fail in only one of `n` distinct ways, because the bundling of the inner `Either` with its `Left` case fixed to `Throwable` prevents such descriptions.

Now, another way to formulate the Scalaz 8 IO bifunctor is as `EitherT[Unexceptional, E, A]`. Arguably, this is a better design, practical considerations aside. However, monad transformers are a terrible fit for Scala, and can easily destroy performance and limit applications of effect monads. Instead of taking this approach, Scalaz 8 IO bakes in the `EitherT` so it has no runtime overhead.

Reply:
> considering an IO monad that never catch exceptions

Is such an IO even useful? IO is meant for those kinds of things.

> the difference is better performance and ease of use (no/less monad transformers).

Then it is just another form of `EitherT[IO, E, A]`? That's a bit confusing, does EitherT have any meaning at that point? Should it be removed?

> See partialSnyc. The strategy here is to deffer constraints on type parameters until you actually need them, to maximize possible usages. It is similar to how one does not need an Order[A] to construct a ISet.empty[A].

I get that, but the idea of turning `Throwable => E` sounds like "My error ADT `E` will always need a `SomeThrowable(t: Throwable)` case". Which *feel* wrong.

Reply:
@jdegoes 

> However, monad transformers are a terrible fit for Scala, and can easily destroy performance and limit applications of effect monads. Instead of taking this approach, Scalaz 8 IO bakes in the EitherT so it has no runtime overhead.

That's actually fair (better performance), even though it makes it bit confusing. Maybe make a mention about EitherT on the scaladoc? I can see other people becoming confused about its relationship with EitherT.

Reply:
@jbgi Ah, I see what you mean now.